crates/sui-pg-db/src/db.rs

```diff
@@ -0,0 +1,547 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::ops::{Deref, DerefMut};
+use std::time::Duration;
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use chrono::NaiveDateTime;
+use diesel::migration::{MigrationSource, MigrationVersion};
+use diesel::pg::Pg;
+use diesel::prelude::*;
+use diesel::sql_types::BigInt;
+use diesel::ExpressionMethods;
+use diesel::OptionalExtension;
+use diesel_async::async_connection_wrapper::AsyncConnectionWrapper;
+use diesel_async::{
+    pooled_connection::{
+        bb8::{Pool, PooledConnection},
+        AsyncDieselConnectionManager,
+    },
+    AsyncConnection, AsyncPgConnection, RunQueryDsl,
+};
+use scoped_futures::ScopedBoxFuture;
+use sui_indexer_alt_framework_store_traits::{
+    CommitterWatermark, Connection as StoreConnection, PrunerWatermark, ReaderWatermark, Store,
+    TransactionalStore,
+};
+use sui_sql_macro::sql;
+use tracing::info;
+use url::Url;
+
+use crate::schema::watermarks;
+use crate::FieldCount;
+
+#[derive(clap::Args, Debug, Clone)]
+pub struct DbArgs {
+    /// Number of connections to keep in the pool.
+    #[arg(long, default_value_t = Self::default().db_connection_pool_size)]
+    pub db_connection_pool_size: u32,
+
+    /// Time spent waiting for a connection from the pool to become available, in milliseconds.
+    #[arg(long, default_value_t = Self::default().connection_timeout_ms)]
+    pub connection_timeout_ms: u64,
+}
+
+/// Wrapper struct over a diesel async connection pool.
+#[derive(Clone)]
+pub struct Db {
+    read_only: bool,
+    pool: Pool<AsyncPgConnection>,
+}
+
+pub type ManagedConnection = AsyncPgConnection;
+/// Type alias for a connection from the pool.
+pub type Connection<'p> = PooledConnection<'p, ManagedConnection>;
+
+/// Wrapper struct over the `Connection` type alias for dealing with the `Store` trait.
+pub struct DbConnection<'a>(Connection<'a>);
+
+#[derive(Insertable, Selectable, Queryable, Debug, Clone, FieldCount)]
+#[diesel(table_name = watermarks)]
+pub struct StoredWatermark {
```

Comment:
As mentioned in our chat earlier today, it might be worth splitting this big module into multiple ones, the original db implementation + tests and the trait implementations. Besides readability, I think it separates the logic better too, in the sense that someone can use the original sui-pg-db implementation purely for interfacing with a postgres db without also using the indexing framework. So I figure it's better to separate out the parts specific to the indexing framework.